game.py

```diff
@@ -1,41 +1,82 @@
-from random import randint
+from operator import itemgetter
+
+from random import choice, randint
 
 from exceptions import NimException
 
 
 class Bot():
+    subclasses = []
+
+    def __init_subclass__(cls, **kwargs):
+        super().__init_subclass__(**kwargs)
+        Bot.subclasses.append(cls)
```

Comment:
@romainrichard, how do you feel about making these real attribute rather than a list?
`setattr(Bot, cls.__name__, cls)`
But it might be a little tricky to grab a random one...
```
this_bot = choice(['foobot', 'barbot'])
getattr(Bot, this_bot).move_strategy(state)

```

Reply:
I don't see what problem that would solve :/

Reply:
I like this better the more I think about it. It loses lines in both the `Bot` and `Chaos` classes, it defines the list you're choosing between in the same place that you are choosing from it without need to filter out the `Chaos` bot from itself, we could even just add `Bot` to `Nim` rather than adding all bots individually. Seems more versatile.

Reply:
You'll have to show some code, can't picture it.

Reply:
```
class Bot():
    def __init_subclass__(cls, **kwargs):
        super().__init_subclass__(**kwargs)
        setattr(Bot, cls.__name__, cls)

    def nonzero_pile_indexes(self, state):
        return [i for i, nb_stones in enumerate(state) if nb_stones >= 1]

    def move(self, state):
        if not self.nonzero_pile_indexes(state):
            raise NimException(f'No valid moves on state: {state}')
        return self.move_strategy(state)


class Chaos(Bot):
    def move_strategy(self, state):
        chosen_bot = choice(['Greedy', 'Random', 'Garbage'])
        return getattr(Bot, chosen_bot).move_strategy(state)
```

Reply:
Then we get to define which bots `Chaos` chooses between, so it won't throw an error when it tries to invoke the `move_strategy` of an unimplemented bot.

Reply:
So we have to manually keep track of the list of bots instead of letting the code to it for us, and we gain what?

Reply:
We gain getting to chose what bot's `Chaos` can chose between. We gain getting to tell it to not use stubbed out bots. We lose line count. We gain being able to just add `Bot` to `Nim` rather than having to add a line for every new bot. We gain that the most important thing gets autoupdated now. (`Nim` gets autoupdated rather than `Chaos`)

Reply:
So every time we create a new `Bot`, we'll need to manually update the list for `Chaos`. I can see us getting to dozens if not hundreds of them, this doesn't seem wise.

> We gain being able to just add `Bot` to `Nim` rather than having to add a line for every new bot.
While this is a problem that we need to solve, I don't think this is the right solution.

I really don't see any advantage to this.

Reply:
We have to update the chaos list, but now we wouldn't need to update the way more important part. All of this goes away...
```
        self.perfect = Perfect()
        self.random = Random()
        self.ai = AI()
        self.garbage = Garbage()
```
And we just do...
```
self.bots = Bot()
```

Reply:
I agree that this needs to be solved, but:
1. We shouldn't instantiate the bots in the first place, we're wasting memory. We should keep track of classes.
2. We have multiple bots, it makes no sense to have each stored as attributes based on their class name. They should be in a list.
3. If I create a `Bot` with a class named `move` or `valid_pile_indexes` for example, this is going to get weird.

Yes we should simplify how we keep track of bots, but this is not the right way.

Reply:
Fair enough. And we should keep it the same just for the sake of getting this PR merged. Ship it.